File: cli/src/labels.ts
import type { DevcontainerConfig } from './types.js';

export const LABEL_NAMESPACE = 'dev.devcontainer-installer';
export const LABEL_MANAGED = `${LABEL_NAMESPACE}.managed`;
export const LABEL_PROJECT = `${LABEL_NAMESPACE}.project`;
export const LABEL_VERSION = `${LABEL_NAMESPACE}.version`;

export const SCHEMA_VERSION = '1';

export function projectId(config: DevcontainerConfig): string {
  return config.image.replace(/[:@/]/g, '_');
}

export function dockerfileLabelBlock(config: DevcontainerConfig): string {
  const project = projectId(config);
  return [
    '# Standard labels — used by the CLI to identify, purge and update images.',
    `LABEL ${LABEL_MANAGED}="true" \\`,
    `      ${LABEL_PROJECT}="${project}" \\`,
    `      ${LABEL_VERSION}="${SCHEMA_VERSION}"`,
  ].join('\n');
}

export function composeLabels(config: DevcontainerConfig): Record<string, string> {
  return {
    [LABEL_MANAGED]: 'true',
    [LABEL_PROJECT]: projectId(config),
    [LABEL_VERSION]: SCHEMA_VERSION,
  };
}

File: cli/internal/domain/types/labels.go
package types

import (
	"fmt"
	"strings"
)

const LabelNamespace = "dev.devcontainer-installer"
const LabelManaged = LabelNamespace + ".managed"
const LabelProject = LabelNamespace + ".project"
const LabelVersion = LabelNamespace + ".version"
const LabelQuickRun = LabelNamespace + ".quick-run"

// LabelSharedConfig marks the single shared tool-config volume so prune can
// identify (and protect) it. The volume name is the primary filter; this label
// is a belt-and-braces identifier applied at `docker volume create`.
const LabelSharedConfig = LabelNamespace + ".shared-config"
const SchemaVersion = "1"

// ImageNamespace is the repository prefix for locally built images
// (devcontainer-cli/<fingerprint>:latest). Prune detects images by the
// managed label instead, so both local and remote images are covered.
const ImageNamespace = "devcontainer-cli"

func ProjectID(config *DevcontainerConfig) string {
	replacer := strings.NewReplacer(":", "_", "/", "_", "@", "_")
	return replacer.Replace(config.Image)
}

func DockerfileLabelBlock(config *DevcontainerConfig) string {
	project := ProjectID(config)
	return fmt.Sprintf(
		"LABEL %s=\"true\" \\\n      %s=\"%s\" \\\n      %s=\"%s\"",
		LabelManaged,
		LabelProject, project,
		LabelVersion, SchemaVersion,
	)
}

func ComposeLabels(config *DevcontainerConfig) map[string]string {
	return map[string]string{
		LabelManaged: "true",
		LabelProject: ProjectID(config),
		LabelVersion: SchemaVersion,
	}
}
